### ISSUE TITLE
EXOMalwareFilterPolicy  -Problem with Auth. via AppReg with Cert

### ISSUE THREAD

**Issue by T630Dev** [microsoft/Microsoft365DSC#1513]
#### Details of the scenario you tried and the problem that is occurring
In the M365DSC version 1.21.1013.1, I am having auth issues for DSC Res that use the App Registration such as EXOMalwareFilterPolicy.
The error claims that the certificate that is used for the App Reg cannot be found or is expired.
I double checked and the certificate is in the machine MY store and is not expired.
Also in the Azure Portal for the target tenant, the certificate is there and valid.

I can manually do a connect-mgGraph which works fine by using certificateThumbprint, TenantID and ClientID parameters.
It does also not help when I run Connect-MgGraph before Start-DSCConfiguration

When I do not use the AppReg but instead only the Credential Attribute it works fine.

#### Verbose logs showing the problem

![image](https://user-images.githubusercontent.com/88185985/138253333-493be963-1128-4054-a35d-12e67a12ecd6.png)
![Expires](https://user-images.githubusercontent.com/88185985/138256861-2dcd0177-62b5-4ff7-a6f3-a6b795bcef61.jpg)

#### The DSC configuration that is used to reproduce the issue (as detailed as possible)
DSC Config file content:
instance of MSFT_EXOMalwareFilterPolicy as $MSFT_EXOMalwareFilterPolicy1ref
{
ResourceID = "[EXOMalwareFilterPolicy]20ae5c1b-9001-47d5-8459-17b503eec607-EXOMalwareFilterPolicy";
 Action = "DeleteAttachmentAndUseDefaultAlert";
 CertificateThumbprint = "AFCC2BED0FF590DD012992EFD707642BB3312";
 Ensure = "Present";
 ApplicationId = "MyAppID...";
 SourceInfo = "::10::3::EXOMalwareFilterPolicy";
 TenantId = "MyTenant.onmicrosoft.com";
 Identity = "Global";
 ModuleName = "Microsoft365DSC";
 ModuleVersion = "1.21.1013.1";

 ConfigurationName = "MainConfig";

};

#### The operating system the target node is running
OsName               : Microsoft Windows Server 2019 Datacenter
OsOperatingSystemSKU : DatacenterServerEdition
OsArchitecture       : 64-Bit
WindowsVersion       : 1809
WindowsBuildLabEx    : 17763.1.amd64fre.rs5_release.180914-1434
OsLanguage           : en-US
OsMuiLanguages       : {en-US, de-DE}

**Comment by NikCharlebois** [comment]
I am able to replicate this behavior whenever the Private key .PFX is not installed in the Local System user's certificate store. Make sure the .PFX gets installed and then try again. The screenshot above shows that the certificate was installed, which is also required but you will also need the private key.

**Comment by T630Dev** [comment]
ok thanks. But that means that you cannot replicate this if the certificate and Prrivate key are installed correctly?
From certlm.msc, I can see that the private key is installed. (see screenshot)
![Private](https://user-images.githubusercontent.com/88185985/139840800-116b6455-3705-4203-aebe-96967f3aeaa1.jpg)

**Comment by NikCharlebois** [comment]
If you run a new PowerShell console as Local System, are you getting the same error when running any of the following commands? Run each one in a separate console.

Connect-M365Tenant -Workload ExchangeOnline -ApplicationId <AppId> -TenantId <TenantId> -CertificateThumbprint <Thumbprint>

or

Connect-ExchangeOnline -AppId <AppId> -OrganizationName <TenantId> -CertificateThumbprint <Thumbprint>

**Comment by T630Dev** [comment]
Connect-ExchangeOnline works fine, except that the parameter is called Organization instead of OrganizationName.
Also Connect-MpGraph works and Export-M365DSCConfiguration with, for example, -Components @("EXOMailboxSettings"), ApplicationID, TenantID, CertificateThumbprint works fine.
I only get the issue when running Start-DSCConfiguration -useExisting after I published my partial config.

**Comment by NikCharlebois** [comment]
After running Connect-M365TenantConfig and getting the error, can you confirm that the information contained in the Authentication cache is correct? 

Simply print out the content of $Global:MSCloudLoginConnectionProfile.MicrosoftGraph

**Comment by T630Dev** [comment]
Looks ok to me. At least the app reg GUID, the tenant name and the cert thumbprint are ok.
![t](https://user-images.githubusercontent.com/88185985/139848943-74f27708-0e9d-43b2-aad8-e3c5d6e08c68.jpg)

**Comment by NikCharlebois** [comment]
Ok then, let's slowly step through the auth stack. Run a new PowerShell console as Local System and run:

Connect-MgGraph -ClientId <AppId> -TenantId <TenantId> -CertificateThumbprint <Thumbprint>

Logic says you should be getting the error then.

**Comment by T630Dev** [comment]
Correct. I get the error as local system but not as my user.
Error message: certificate was not found or has expired.
Even when I assign Full access to everyone to the private key of the cert, it does not resolve this.

**Comment by NikCharlebois** [comment]
Am I misreading this or did the certificate expired back in July? (based on your earlier screenshot)

**Comment by T630Dev** [comment]
Yes, it expires in July in 100 years:)

**Comment by NikCharlebois** [comment]
Ok, let's see if we can replicate it on my end. 
For me, the only certificate I have installed for Local System is the certificate I am using for M365DSC and the one I uploaded for my Azure AD App Registration:
![image](https://user-images.githubusercontent.com/2547149/139862961-c2a01bf0-392a-459d-9b11-a500734fed02.png)

Running connect-mggraph with it works as expected when running as local system:
![image](https://user-images.githubusercontent.com/2547149/139863234-3e2d6eaf-f0e0-48a6-84c6-5f44414f4407.png)

the moment I uninstall that certificate from the local system's store, I get the same error as you:
![image](https://user-images.githubusercontent.com/2547149/139863381-3d8b64da-0631-4184-a72f-b2288f418c40.png)

This proves at least one thing, the issue is with the cert and not M365DSC. Now, to get to the source of the issue, we need to figure out why Local System can't connect to the out-of-the-box Connect-MgGraph cmdlet.

**Comment by T630Dev** [comment]
Yeah that would be interesting. I can generate a new cert in the Key Vault and check again.
Any specific settings you set in your cert?
Did you use a self signed cert? Did you also put it in the trusted root CA store? 
Since I am using an Windows Server 2019 VM in Azure, there also no special config / GPOs there.

**Comment by NikCharlebois** [comment]
It is **not** in the Trusted root.
I am not aware of any specific GPOs that could be causing this nor have I ever encountered this.
I am using a self-signed cert and this is normally how I generate them:

New-SelfSignedCertificate -Type DocumentEncryptionCertLegacyCsp `
            -DnsName 'Microsoft365DSC' `
            -Subject 'M365DSCEncryptionCert' `
            -HashAlgorithm SHA256 `
            -NotAfter (Get-Date).AddYears(10)

**Comment by T630Dev** [comment]
I reviewed the Connect-MgGraph parameters and it seems that by default the command is not searching the certificate in the local machine store but in the current user store. If I then connect as local system, the certificate is expected to be in the current user store of the local system. That's why it cannot find it.

As a workaround, the certificate parameter lets you pass a X509Certificate2 object. I tried that and also imported the PFX file into the current user store of the local system. In both cases, I get the following error:

![Error](https://user-images.githubusercontent.com/88185985/139874373-33fcfddd-007e-496b-b142-00ae9083cd94.jpg)
Tried then to convert the thumbprint to bytes and then to base 64 string but that did not work either.
